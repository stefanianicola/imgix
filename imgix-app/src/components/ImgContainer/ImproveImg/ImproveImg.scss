@import "../../../sass/all";
.improves-content {
    background-color: #0099cb;
    display: flex;

    .buttons {
        margin: 0;
        display: flex;
        padding-top: 10px;
        padding-left: 10px;
        &:last-child {
            padding-bottom: 30px;
        }
        @include respond-to(medium) {
            padding-left: 30px;
        }
        .btn {
            padding: 0;
            font-size: 13px;
            font-weight: 100;
            color: #fff;
            text-decoration: none;
            @include respond-to(small) {
                font-size: 18px;
            }
            &:focus {
                box-shadow: none;
            }
        }
    }
}
