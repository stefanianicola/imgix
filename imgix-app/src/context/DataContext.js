import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios';
import Params from '../Data/Params';
export const DataContext = createContext();


const DataContextProvider = (props) => {
    const [data, setData] = useState([]);
    const [flag, setFlag] = useState(false);
    const [btn, setBtn] = useState([]);
    const [nameButtons, setNameButtons] = useState([]);
    const [loading, setLoading] = useState(false);
    const [params, setParams] = useState([]);
    const [url, setUrl] = useState("");
    const [hideMenu, setHideMenu] = useState(false);

    const urlImg = "https://assets.imgix.net/unsplash/";

    //consume photos API
    useEffect(() => {
        setLoading(true);
        try {
            axios("https://storage.googleapis.com/nanlabs-engineering-technical-interviews/imgix-samples-list.json")
                .then((res) => {
                    setData(res.data);
                })
        } catch (error) {
            console.log(error);
        }
        setLoading(false);
    }, [])

    //Pass data from the params to the state of btn so that the names appear in the buttons
    useEffect(() => {
        setNameButtons(Params);
    }, [])


    //Pass the parameters of the button that you click
    const addControl = (v) => {
        let parametros = [...Params]
        parametros.forEach((param) => {
            if (v === param.alias) {
                setBtn(param)
            }
        })
        setFlag(true);
    }

    // filter the data of the params to be able to take a single value per parameter
    // and build the url with all the effects we want in the photos
    const addParamUrl = (alias, value) => {
        let aux = [...params];
        let isParam = false;

        aux.forEach((a) => {
            if (a.alias === alias) {
                isParam = true;
                a.value = value;
                setParams(aux);
            }
        })
        !isParam &&
            setParams([...aux, { alias, value }]);
        convertUrl();

    }

    //Convert the array of objects into a string to use as the url of the image
    const convertUrl = () => {
        let stringP = "";
        for (let i = 0; i < params.length; i++) {
            stringP = stringP + `${params[i].alias}=${params[i].value}&`

        }
        setUrl(stringP);

    }

    //reset url with change img
    const resetUrl = () => {
        setUrl("");
        setParams([]);
        setFlag(false);
    }

    //delete the last parameter added
    const removeItem = () => {
        params.pop();
        convertUrl();
        setFlag(false);
    }

    return (
        <DataContext.Provider
            value={{
                data,
                setData,
                addControl,
                nameButtons,
                flag,
                btn,
                loading,
                params,
                setParams,
                addParamUrl,
                url,
                removeItem,
                hideMenu,
                setHideMenu,
                resetUrl,
                urlImg
            }}>
            {props.children}
        </DataContext.Provider>
    )
}

export default DataContextProvider;
