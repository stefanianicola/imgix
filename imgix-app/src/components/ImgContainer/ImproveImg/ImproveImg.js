import React, { useContext } from 'react';
import './ImproveImg.scss';
import { Button } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import { DataContext } from '../../../context/DataContext';

const ImproveImg = (props) => {
    const { addControl } = useContext(DataContext);

    return (
        <div className="buttons">
            <Button variant="link" className="d-block ml-2" onClick={() => addControl(props.alias, props.min, props.max)}>{props.name}</Button>
        </div>
    )
}
export default ImproveImg;