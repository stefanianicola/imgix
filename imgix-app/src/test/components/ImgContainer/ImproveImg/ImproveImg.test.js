import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import ImproveImg from '../../../../components/ImgContainer/ImproveImg/ImproveImg'
import { DataContext } from '../../../../context/DataContext';
import { ButtonToolbar } from 'react-bootstrap';

describe('ImproveImg component', () => {
    const addControl = jest.fn();
    const params = {
        "id": 11,
        "name": "Unsharp Mask",
        "alias": "usm",
        "value": 0,
        "min": -100,
        "max": 100
    };

    const wrapper = mount(
        <Router>
            <DataContext.Provider value={{ addControl }} >
                <ImproveImg props={params} />
            </DataContext.Provider>
        </Router>
    );


    it('should be create the component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should eject addControl function', () => {
        const btn = wrapper.find('button');
        btn.simulate('click');

        expect(addControl).toHaveBeenCalled();
    });

})