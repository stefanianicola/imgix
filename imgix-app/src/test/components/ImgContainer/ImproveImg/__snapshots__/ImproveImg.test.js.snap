// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImproveImg component should be create the component 1`] = `
<div
  className="buttons"
>
  <button
    className="d-block ml-2 btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  />
</div>
`;
